### PR TITLE
refactor(transformer): replace `&’a Trivias` with `Rc<Trivias>`

### PR DESCRIPTION
**Pull request by Dunqing**
`Transformer` needs to borrow `Trivias`. https://github.com/rolldown/rolldown/pull/1364/files/8be1cc80520e1a466cd7b2d3fb8b637db9e529a9#r1630711060

### CODE REVIEW

**Comment by Dunqing** [comment]
* **#3580** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3580?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by Boshen** [comment]
No we should never clone this heavy data structure, we need to `Rc` or `Arc` it before hand.

**Comment by overlookmotel** [comment]
I also ran into problems with trying to share `Trivias` in another context. It's problematic that when you create a reference `&Trivias`, its lifetime doesn't live as long as `Program<'a>`.

I'm not sure `Rc`/`Arc` will solve this problem as it'll still not live as long as we need it to.

In my view the best solution would be to store `Trivias` in the allocator, and hand out a `TriviasRef<'a>` with same lifetime as `Program<'a>`. Then it'd be easy to send both around the place together.

Ideally the actual `BTreeMap` data would be stored in an arena, so it can be cheap to drop, but I can't find a `BTreeMap` implementation which supports arena allocators. So for now we could just add it as a field on `Allocator`. `Allocator` owns `Trivias`, and it'd get dropped when allocator is dropped.

Would that solve the problem that this PR is targeting?

**Comment by Dunqing** [comment]
> Would that solve the problem that this PR is targeting?

I think would solve this problem. I will give it a try!

**Comment by overlookmotel** [comment]
OK great. We should also add a function to `Allocator` that resets both the bumpalo arena and `Trivias`:

```rs
impl Allocator {
  fn reset(&mut self) {
    self.bump.reset();
    self.trivias.reset();
  }
}

impl Trivias {
  fn reset(&mut self) {
    self.comments.clear();
    self.irregular_whitespaces.clear();
  }
}
```

Otherwise `Trivias` will keep growing each time the allocator is reset.

**Comment by Dunqing** [comment]
I tried to give `Trivias` the same lifetime as `Program`. However, this approach didn't solve this problem. The `Transformer` always needs to borrow the `Trivias`. So we cannot move out `ast` at the same time.
image

In my opinion, the only viable solution is to make `Trivias` cloneable. By using `Rc`, we can address this problem. I found the same solution [here](https://github.com/oxc-project/oxc/blob/532510d7143b1c6010cf078cd158a8af17bc982d/crates/oxc_semantic/src/builder.rs#L110-L115)

I don't know much about lifetime, so maybe there's a better solution I don't know about.

**Comment by overlookmotel** [comment]
Oh dear! I'm not good with lifetimes either. I don't think any of us are!

If you can solve it with `Rc`, probably best to do that for now to unblock us. I don't think it's the optimal solution, but we can come back to it later if we want to try to remove `Rc`. Reasoning about the lifetimes in transformer is a brain-bender.

File: crates/oxc_wasm/src/lib.rs [hunk 1]
```diff
@@ -170,16 +170,17 @@ impl Oxc {
         let ret = Parser::new(&allocator, source_text, source_type)
             .allow_return_outside_function(parser_options.allow_return_outside_function)
             .parse();
+        let trivias = Rc::new(ret.trivias);
 
-        self.comments = self.map_comments(&ret.trivias);
+        self.comments = self.map_comments(&trivias);
```
Comment (overlookmotel) [inline crates/oxc_wasm/src/lib.rs:175]: I think you can leave this as `&ret.trivias` (`map_comments` only needs to temporarily borrow `trivias`). Then wrap `trivias` in an `Rc` further down:

```rs
let semantic_ret = SemanticBuilder::new(source_text, source_type)
    .with_trivias(Rc::new(&ret.trivias))
```

That'd remove one `Rc::clone`.

File: crates/oxc_semantic/src/builder.rs [hunk 2]
```diff
@@ -108,8 +108,8 @@ impl<'a> SemanticBuilder<'a> {
     }
 
     #[must_use]
-    pub fn with_trivias(mut self, trivias: Trivias) -> Self {
-        self.trivias = Rc::new(trivias);
+    pub fn with_trivias(mut self, trivias: Rc<Trivias>) -> Self {
+        self.trivias = trivias;
         self.jsdoc = JSDocBuilder::new(self.source_text, &self.trivias);
```
Comment (overlookmotel) [inline crates/oxc_semantic/src/builder.rs:113]: Please see #3586.

File: crates/oxc_wasm/src/lib.rs (hunk 1 above)
Comment (Dunqing) [reply crates/oxc_wasm/src/lib.rs]: Yes!

File: crates/oxc_wasm/src/lib.rs [hunk 3]
```diff
@@ -234,7 +236,7 @@ impl Oxc {
                 &path,
                 source_type,
                 source_text,
-                semantic.trivias(),
+                Rc::clone(&trivias),
```
Comment (overlookmotel) [inline crates/oxc_wasm/src/lib.rs:239]: I think this can just be `trivias`. No need to clone the `Rc`, as it's not used again after this.

**Comment by overlookmotel** [comment]
I'd suggest making the change above, but it's not so important.

Besides that, this change is motivated by a need in Rolldown, and I'm not familiar enough with Rolldown to say "LGTM", as I don't understand the context. But if it solves the problem, everything else looks good to me.

File: crates/oxc_wasm/src/lib.rs (hunk 3 above)
Comment (Dunqing) [reply crates/oxc_wasm/src/lib.rs]: Thanks for catching that

**Comment by Dunqing** [comment]
> Besides that, this change is motivated by a need in Rolldown, and I'm not familiar enough with Rolldown to say "LGTM", as I don't understand the context. But if it solves the problem, everything else looks good to me.

Thank you for reviewing this! I tested this change in Rolldown and it worked.

**Comment by overlookmotel** [comment]
> Thank you for reviewing this! I tested this change in Rolldown and it worked.

OK great. Merge it then I guess!

**Pull request merged**